### PR TITLE
Enable Update

### LINKED ISSUE #293: ETABS_Toolkit: Update Material functionality
#### Description:

Add method for updating a material in Etabs

### LINKED ISSUE #294: ETABS_Toolkit: Add Update SectionProperty functionality
#### Description:

Add update method for SectionProeprty

### LINKED ISSUE #295: ETABS_Toolkit: Add update method for SurfaceProperties

### LINKED ISSUE #296: ETABS_Toolkit: Add update method for Bar

### PR DESCRIPTION
**Pull request by kThorsager**
### Issues addressed by this PR

Closes #293 
Closes #294 
Closes #295 
Closes #296 

Adds update methods which can be called by the Push for the main elements which can be pushed to ETABS.

Also noticed that ETABS has a separate namespace `cEditFrame` which contains one method to change a frames connectivity, which resolves the issue about AddByPoint sometimes taking liberties with the order of the nodes. And as the existing solution was essentially a hack around that (pre-flipping our Bars) I removed that.

 ### Test files

https://burohappold.sharepoint.com/:u:/s/BHoM/EcYeRgiNXA9InJ4fh1WyJB4B05hhc13F4NK29_9aJLoWCQ?e=BJagys

 ### Changelog
- Update functionallity for `Material`, `SectionProperty`, SurfaceProperty` & `Bar`
- Changed API workaround for `Bar`s to a more stable one

 ### Additional comments

A note is that frame properties, technically aren't reinitialised, but it looks like all of their properties are included in the set method, so you know, there's not much difference.

But tests on Surface properties and Materials imply that it's true, (as it works there)

### CODE REVIEW

**Comment by kThorsager** [comment]
Or well, everything was less fun than expected as ETABS16 does not have the Change connectivity method. (which does explain why I didn't find it last time)

**Comment by kThorsager** [comment]
> - Changed API workaround for Bars to a more stable one

So that was a modified truth.
the ChangeConnectivity method is also strange in many ways, but only when the object is moved (?) so for Push purposes it's no problem.
And added another janky solution for the Update Bar case

File: Etabs_Adapter/CRUD/Update/Bar.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+#if Debug16 || Release16
+                string start = "";
+                string end = "";
+                m_model.FrameObj.GetPoints(name, ref start, ref end);
+
+                if (bhBar.StartNode.CustomData[AdapterIdName].ToString() != start ||
+                    bhBar.EndNode.CustomData[AdapterIdName].ToString() != end)
+                {
+                    Engine.Reflection.Compute.RecordWarning("ETABS16 does not support Update of Bar connections.");
```
Comment (IsakNaslundBh) [inline Etabs_Adapter/CRUD/Update/Bar.cs:93]:
```suggestion
                    Engine.Reflection.Compute.RecordWarning("ETABS16 does not support Update of Bar connectivity, which means the geometry can not be updated. To update the connectivity or position of a Bar,  delete the existing Bar you want to update and create a new one.");
```

File: Etabs_Adapter/CRUD/Update/Bar.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+            }
+
+#if Debug16 || Release16
+            // Not a problem in 2016 since it can't move them at all.
+#else
+            // ETABS refuses to update the location of Bars unless a edit command for them is run.
+            // Mainly a UI bug, but seem to contain other artifacts as well.
+            // Does not count as a command if the movment is zero.
+            m_model.SelectObj.All();
```
[the remark this answers is not in the document]
Comment (kThorsager) [reply Etabs_Adapter/CRUD/Update/Bar.cs:108]: I have tested that and closing and opening the model, and that does not work either. So not sure what goes on behind the scenes

File: Etabs_Adapter/CRUD/Update/Material.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    
+                    success &= SetObject(material);
+                }
+                else
+                {
+                    // No material of that name found
+                    Engine.Reflection.Compute.RecordWarning($"Failed to update material: {material.DescriptionOrName()}, as no such material was present in the model.");
+                }
+
+                if (!success)
```
[the remark this answers is not in the document]
Comment (kThorsager) [reply Etabs_Adapter/CRUD/Update/Material.cs:86]: yeah, but I set it to true at the beginning of each loop, what is weird is that I'm returning the success for the last material only

File: Etabs_Adapter/CRUD/Update/Bar.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Etabs_Adapter/CRUD/Update/Bar.cs:109]: fair enough. Lets stick with this then as it seem to at least work

**Pull request merged**